### DISCUSSION TITLE
Feedback on the Survey-Text Type Trial

### DISCUSSION THREAD

**Discussion by dusannikolic92** [jspsych/jsPsych#2396] [Q&A] [answered]
Hello,

I am Dusan and I am trying to develop a simple sequence recall test.
I am trying to develop a simple feedback (correct, incorrect) on the answers provided in the text field by my participants. My participants need to, based on what they hear, insert keys 'a' and 'l' in the correct order. E.g. they hear a sequence and they insert into the text field a sequence 'ala'. Based on this text sequence, I want to have a feedback, but it does not work with survey-text plugin because survey-text plugin is storing the data in some way unbeknownst to me. I have tried numerous combinations of what goes into the timeline variable 'correct_response'. I have tried with "a", with ['a'], with what you can see below, etc. but nothing works. Even though the data.correct_response and data.response are identical in the code I provide below, which I could check in Cognition.run, the code still won't recognize them as identical. 
I would appreciate any help on this! I have tried with JSON.stringify as well, but it seems that I am doing something wrong. If you could provide an example of how JSON.stringify would work if you are using this function, that would be great. Thank you!

```javascript
var timeline = [];
var repeat_prac = false; 
  
var practice_timeline = [
{ 
    correct_response: {"Q0":"a"},
},
{ 
    correct_response: {"Q0":"a"},
},
{ 
    correct_response: {"Q0":"a"},
},
{
    correct_response: {"Q0":"a"},
},
{
    correct_response: {"Q0":"a"},
},
{
    correct_response: {"Q0":"a"},
}
];


var instructions = {
  type: 'html-button-response',
  stimulus: 'Something',
  choices: ['Next']
};


var practice_trial = {
  type: 'survey-text',
   questions: [
          {prompt: 'Answer', required: true},
              ],
   data: {
              correct_response: jsPsych.timelineVariable('correct_response'),
              trial_part: 'practice'
   },
            on_finish: function(data) {

                var acc = false;
                if (jsPsych.pluginAPI.compareKeys(data.correct_response, data.response)) {
                    acc = true;
                }
               data.accuracy = acc;
            }
        };
        
  
        var feedback = {
            type: 'html-keyboard-response',
            stimulus: function() {
                var feedback_text = '<span style="color:red;font-size:30px;">Incorrect.</span>';
                var last_resp_acc = jsPsych.data.getLastTrialData().values()[0].accuracy;
                if (last_resp_acc == true) {
                    feedback_text = '<span style="color:green;font-size:30px;">Correct!</span>'
                }
                return feedback_text;
            },
            choices: jsPsych.NO_KEYS,
            trial_duration: 2000
        };

        var prac_procedure = {
            timeline: [practice_trial, feedback],
            timeline_variables: practice_timeline,
            randomize_order: true
        };
        

var repeat_prac_message = {
        type: 'html-button-response',
        stimulus: '<p>Not enough correct responses.</p>'+
                  '<p>Repeat everything.</p>',
        choices: ['Yes, please']
    };

    var repeat_prac_conditional = {
        timeline: [repeat_prac_message],
        conditional_function: function() {
            var last_prac_trials = jsPsych.data.get().filter({trial_part:'practice'}).last(practice_trial.length);
            var n_correct = last_prac_trials.filter({accuracy: true}).count();
            var prop_corr = n_correct/practice_trial.length;
            if (prop_corr < 0.6) {
                repeat_prac = true;
                return true;
            } else {
                repeat_prac = false;
                return false;
            }
        }
    }

var instructions_prac_loop = {
            timeline: [instructions, prac_procedure, repeat_prac_conditional],
            loop_function: function() {
                if (repeat_prac == true) {
                    return true;
                } else {
                    return false;
                }
            }
        }
        timeline.push(instructions_prac_loop); 

  /* start the experiment */
    var block = {
    type: 'html-keyboard-response',
    choices: ['n'], 
    stimulus: '<p>Press "a" to complete the experiment.</p>',
    stimulus_width: 300,
    on_finish: function(data) {
      if (jsPsych.pluginAPI.compareKeys(data.response, 'a')) {
        jsPsych.endExperiment('Success. You can exit.');
      }
    }
        };
   timeline.push(block);
  
    
        jsPsych.init({
            timeline: timeline,
            show_preload_progress_bar: true,
            message_progress_bar: "Progress",
            show_progress_bar: true,
            use_webaudio: false,
            case_sensitive_responses: false,
        }) 
```

**Comment by becky-gilbert** [comment, marked as answer]
Hi @dusannicolik92, I'm guessing you're using jsPsych v6.3? Do you see any errors in the console? It might also be helpful to see the output of `console.log(data)` in your response trial, just to double check the format of the data.

I think the problem might be the use of the `jsPsych.pluginAPI.compareKeys` function here. That function is intended to compare individual key press strings, e.g. 'a', 'l'. (And the only reason this function exists is so that two key presses can be compared in a case sensitive or insensitive way, depending on the experiment settings).

In your code, you're comparing two _objects_, and even if you extract the response string from the objects, there can be more than one character in the response (e.g. `{Q0: 'ala'}`). So I wouldn't use the `compareKeys` function - instead I would extract the two response strings from the objects and compare them directly:
```js
{
  correct_response: 'a'
}

// ...

if (data.response.Q0 == data.correct_response) {
  acc = true;
}
```
And if you want to make sure that the comparison isn't case sensitive, you could call `toLowerCase` on the response string.

Does that work?

BTW a lot of people seem confused about when to use the jsPsych `compareKeys` function, so I think this is something we need to clarify in the documentation.

**Reply by dusannikolic92** [reply]
Hi @becky-gilbert, thank you very much. It works!

I am aware of the comparekeys plugin, and I've read the documentation, and even used it with an AX task, and I know that it compares key strings or numbers, but I could not find any alternative. Perhaps, it can be clarified in the documentation what it is mainly used for. I would never guess to attach the 'Q0' to the data.response, so thank you for that!

Best,
Dusan

**Reply by becky-gilbert** [reply]
Ok great, thanks for the feedback! We'll modify the documentation to explain how to do these other kinds of comparisons 👍
